# 2.4: Evaluating Hyperparameters - Deep Learning CNN Model – Bayseian Model

## Part 2: Deep Learning CNN Model – Bayseian Model

Objective: Optimise hyperparameters for CNN Model.

### This script contains the following:

## 1. [Import Libraries and Data](#library)

## 2. [Data Wrangling](#wrangling)

## 3. [Reshape for modeling](#reshape)

## 4. [Data Split](#split)

## 5. [Bayesian Hypeparameter Optimisation](#bayseian)


# Import Libraries and Data <a id='library'></a>


In [1]:
!pip install scikit-learn==1.4.2
!pip install scikeras
!pip install bayesian-optimization
!pip install tensorflow

import pandas as pd
import numpy as np
import seaborn as sns
import os
import operator
import time
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer, accuracy_score
from sklearn.model_selection import StratifiedKFold
from sklearn.utils.multiclass import type_of_target
import tensorflow as tf
from numpy import unique
from numpy import reshape
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score
from tensorflow.keras.layers import Input, Conv1D, Dense, Dropout, BatchNormalization, Flatten, MaxPooling1D
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam, SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from scikeras.wrappers import KerasClassifier  # Use scikeras for scikit-learn compatibility
from math import floor
from bayes_opt import BayesianOptimization
from tensorflow.keras.layers import LeakyReLU  # Use tensorflow.keras instead of keras
LeakyReLU = LeakyReLU(negative_slope=0.1)
import warnings

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 18.5 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1


In [2]:
# Set option to ensure charts are displayed inline in the notebook

%matplotlib inline

In [3]:
# Set option to display max number of columns

pd.set_option('display.max_columns', None)

In [4]:
path = r'/content/drive/MyDrive/Careerfoundary/Machine_learning'

In [5]:
# Import cleaned weather observations data

df = pd.read_csv(os.path.join(path, '012 Data', 'Prepared_Data', 'weather_cleaned_nodate.csv'), index_col = False)

In [6]:
df.head()

BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  BASEL_global_radiation  \
0                  7            0.85           1.018                    0.32   
1                  6            0.84           1.018                    0.36   
2                  8            0.90           1.018                    0.18   
3                  3            0.92           1.018                    0.58   
4                  6            0.95           1.018                    0.65   

   BASEL_precipitation  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0.09             0.7              6.5             0.8   
1                 1.05             1.1              6.1             3.3   
2                 0.30             0.0              8.5             5.1   
3                 0.00             4.1              6.3             3.8   
4                 0.14             5.4              3.0            -0.7   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0            10.9                     1               0.81             1.0195   
1            10.1                     6               0.84             1.0172   
2             9.9                     6               0.77             1.0179   
3            10.6                     8               0.93             1.0268   
4             6.0                     8               0.99             1.0286   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00                7.0   
1                       0.25                    0.00                0.0   
2                       0.67                    0.00                3.5   
3                       0.25                    0.00                0.0   
4                       0.25                    0.06                0.0   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0                 3.7               -0.9                7.9   
1                 2.9                2.2                4.4   
2                 3.1               -0.5                6.4   
3                 2.0               -2.0                3.0   
4                 2.0                0.7                2.8   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                     4               0.67              1.017   
1                     4               0.67              1.017   
2                     4               0.67              1.017   
3                     4               0.67              1.017   
4                     4               0.67              1.017   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                       0.44                    0.01                2.3   
1                       0.18                    0.31                0.0   
2                       0.30                    0.00                0.6   
3                       0.19                    0.00                0.0   
4                       0.19                    0.00                0.0   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0                 2.4               -0.4                5.1   
1                 2.3                1.4                3.1   
2                 2.7                1.7                5.3   
3                 2.0                0.4                4.4   
4                 2.5                1.1                5.3   

   DEBILT_cloud_cover  DEBILT_humidity  DEBILT_pressure  \
0                   7             0.85           1.0032   
1                   8             0.90           1.0056   
2                   6             0.92           1.0165   
3                   8             0.95           1.0265   
4                   6             0.90           1.0243   

   DEBILT_global_radiation  DEBILT_precipitation  DEBILT_sunshine  \
0                     0.07                  0.25              0.0   
1                     0.14              

In [7]:
df.shape

(22950, 135)

In [8]:
# Import predictions data

df_pleasant = pd.read_csv(os.path.join(path, '012 Data', 'Original_Data', 'Dataset-Answers-Weather_Prediction_Pleasant_Weather.csv'), index_col = False)

In [9]:
df_pleasant

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0      19600101                       0                          0   
1      19600102                       0                          0   
2      19600103                       0                          0   
3      19600104                       0                          0   
4      19600105                       0                          0   
...         ...                     ...                        ...   
22945  20221027                       0                          0   
22946  20221028                       0                          0   
22947  20221029                       0                          0   
22948  20221030                       0                          0   
22949  20221031                       0                          0   

       BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   
...                          ...                      ...   
22945                          0                        0   
22946                          0                        0   
22947                          0                        0   
22948                          0                        0   
22949                          0                        0   

       DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
22945                            0                          0   
22946                            0                          0   
22947                            0                          0   
22948                            0                          0   
22949                            0                          0   

       KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
...                        ...                         ...   
22945                        0                           0   
22946                        0                           0   
22947                        0                           0   
22948                        0                           0   
22949                        0                           0   

       MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   
...                            ...                      ...   
22945                            0                        0   
22946                            0                        0   
22947                            0                        0   
22948                            0                        0   
22949                            0                        0   

       MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                              0                      0   
1                              0    

In [10]:
df_pleasant.shape

(22950, 16)

# 2. Data Wrangling <a id='wrangling'></a>

In [11]:
# Drop unnecessary columns
df.columns


Index(['BASEL_cloud_cover', 'BASEL_humidity', 'BASEL_pressure',
       'BASEL_global_radiation', 'BASEL_precipitation', 'BASEL_sunshine',
       'BASEL_temp_mean', 'BASEL_temp_min', 'BASEL_temp_max',
       'BELGRADE_cloud_cover',
       ...
       'STOCKHOLM_temp_max', 'VALENTIA_cloud_cover', 'VALENTIA_humidity',
       'VALENTIA_pressure', 'VALENTIA_global_radiation',
       'VALENTIA_precipitation', 'VALENTIA_sunshine', 'VALENTIA_temp_mean',
       'VALENTIA_temp_min', 'VALENTIA_temp_max'],
      dtype='object', length=135)

In [12]:
df.columns

Index(['BASEL_cloud_cover', 'BASEL_humidity', 'BASEL_pressure',
       'BASEL_global_radiation', 'BASEL_precipitation', 'BASEL_sunshine',
       'BASEL_temp_mean', 'BASEL_temp_min', 'BASEL_temp_max',
       'BELGRADE_cloud_cover',
       ...
       'STOCKHOLM_temp_max', 'VALENTIA_cloud_cover', 'VALENTIA_humidity',
       'VALENTIA_pressure', 'VALENTIA_global_radiation',
       'VALENTIA_precipitation', 'VALENTIA_sunshine', 'VALENTIA_temp_mean',
       'VALENTIA_temp_min', 'VALENTIA_temp_max'],
      dtype='object', length=135)

In [13]:
df.shape

(22950, 135)

In [14]:
# Drop date column on df_pleasant
df_pleasant.shape


(22950, 16)

In [15]:
df_pleasant.drop(columns = 'DATE', inplace = True)

In [16]:
df_pleasant.shape

(22950, 15)

# Reshaping for modeling <a id='reshape'></a>

The Bayesian optimization requires a slightly different form for your X and y sets. Before you split them:

*  Your X shape will need to be in the form (22950, 15, 9);

*  Your y shape will need to be in the form (22950,);

*   You can use the argmax function shown in the Exercise to transform your y set.




**The Bayesian optimization function only accepts y data in “multiclass” and “binary” layouts but not in “multilabel-indicator.”**

In [17]:
# Turn X and answers from a df to arrays

X = np.array(df)
y = np.array(df_pleasant)

In [18]:
X = X.reshape(-1,15,9)

In [19]:
X.shape

(22950, 15, 9)

In [20]:
# Use argmax to transform y

y =  np.argmax(y, axis = 1)
y

array([0, 0, 0, ..., 0, 0, 0])

In [21]:
y.shape

(22950,)

In [22]:
# Check y layout

from sklearn.utils.multiclass import type_of_target
type_of_target(y)

'multiclass'

# Data Split <a id='split'></a>

In [23]:
# Split data into train and test sets

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)

In [24]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(17212, 15, 9) (17212,)
(5738, 15, 9) (5738,)


# Bayesian Hyperparameter Optimization <a id='bayseian'></a>

In [35]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = 15 # Number of weather stations
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

In [36]:
#Create function
def bay_area(neurons, activation, kernel, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate):
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(learning_rate=learning_rate), 'SGD':SGD(learning_rate=learning_rate),
                 'RMSprop':RMSprop(learning_rate=learning_rate), 'Adadelta':Adadelta(learning_rate=learning_rate),
                 'Adagrad':Adagrad(learning_rate=learning_rate), 'Adamax':Adamax(learning_rate=learning_rate),
                 'Nadam':Nadam(learning_rate=learning_rate), 'Ftrl':Ftrl(learning_rate=learning_rate)}
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']

    neurons = round(neurons)
    kernel = round(kernel)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)

    def cnn_model():
        model = Sequential()
        model.add(Conv1D(neurons, kernel_size=kernel,activation=activation, input_shape=(timesteps, input_dim)))
        #model.add(Conv1D(32, kernel_size=1,activation='relu', input_shape=(timesteps, input_dim)))

        if normalization > 0.5:
            model.add(BatchNormalization())
        for i in range(layers1):
            model.add(Dense(neurons, activation=activation)) #(neurons, activation=activation))
        if dropout > 0.5:
            model.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            model.add(Dense(neurons, activation=activation))
        model.add(MaxPooling1D())
        model.add(Flatten())
        model.add(Dense(n_classes, activation='softmax')) #sigmoid for multi-class but won't work

        # Create a new optimizer instance for each model
        if optimizer == 'Adam':
            opt = Adam(learning_rate=learning_rate)
        elif optimizer == 'SGD':
            opt = SGD(learning_rate=learning_rate)
        elif optimizer == 'RMSprop':
            opt = RMSprop(learning_rate=learning_rate)
        elif optimizer == 'Adadelta':
            opt = Adadelta(learning_rate=learning_rate)
        elif optimizer == 'Adagrad':
            opt = Adagrad(learning_rate=learning_rate)
        elif optimizer == 'Adamax':
            opt = Adamax(learning_rate=learning_rate)
        elif optimizer == 'Nadam':
            opt = Nadam(learning_rate=learning_rate)
        elif optimizer == 'Ftrl':
            opt = Ftrl(learning_rate=learning_rate)

        # model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy']) # multi-class won't work
        # ValueError: Supported target types are: ('binary', 'multiclass'). Got 'multilabel-indicator' instead."
        model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy']) #categorical_crossentropy
        # model.summary()
        return model
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=2, patience=20)
    nn = KerasClassifier(build_fn=cnn_model, epochs=epochs, batch_size=batch_size, verbose=2)
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, X_train, y_train, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    return score


In [38]:
start = time.time()
params ={
    'neurons': (10, 100),
    'kernel': (1, 3),
    'activation':(0, 9),
    'optimizer':(0,7),
    'learning_rate':(0.01, 1),
    'batch_size': (200, 1000), #(10, 50), #
    'epochs':(20, 100),
    'layers1':(1,3),
    'layers2':(1,3),
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}
# Run Bayesian Optimization
nn_opt = BayesianOptimization(bay_area, params, random_state=42)
nn_opt.maximize(init_points=5, n_iter=4) #25
print('Search took %s minutes' % ((time.time() - start)/60))

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  kernel   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:73: FutureWarning: `fit_params` is deprecated and will be removed in version 1.6. Pass parameters via `params` instead.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/32
15/15 - 5s - 357ms/step - accuracy: 0.3684 - loss: 42.6661
Epoch 2/32
15/15 - 6s - 411ms/step - accuracy: 0.4615 - loss: 49.4307
Epoch 3/32
15/15 - 4s - 255ms/step - accuracy: 0.5040 - loss: 29.4223
Epoch 4/32
15/15 - 5s - 343ms/step - accuracy: 0.4850 - loss: 21.1952
Epoch 5/32
15/15 - 6s - 416ms/step - accuracy: 0.5113 - loss: 16.7492
Epoch 6/32
15/15 - 4s - 259ms/step - accuracy: 0.4658 - loss: 19.8701
Epoch 7/32
15/15 - 5s - 340ms/step - accuracy: 0.5010 - loss: 14.3623
Epoch 8/32
15/15 - 5s - 342ms/step - accuracy: 0.5044 - loss: 14.4466
Epoch 9/32
15/15 - 5s - 334ms/step - accuracy: 0.4989 - loss: 12.3193
Epoch 10/32
15/15 - 8s - 526ms/step - accuracy: 0.5300 - loss: 12.8547
Epoch 11/32
15/15 - 8s - 503ms/step - accuracy: 0.5129 - loss: 12.5088
Epoch 12/32
15/15 - 6s - 403ms/step - accuracy: 0.5005 - loss: 12.1745
Epoch 13/32
15/15 - 4s - 276ms/step - accuracy: 0.5329 - loss: 11.2865
Epoch 14/32
15/15 - 6s - 369ms/step - accuracy: 0.5119 - loss: 11.6292
Epoch 15/32
15/

/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/32


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/32


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/32


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
-----------------------------------------------------------

| 1         | nan       | 3.371     | 960.6     | 0.732     | 0.1796    | 32.48     | 1.312     | 1.116     | 2.732     | 0.6051    | 73.73     | 0.02058   | 6.789     |
Epoch 1/44
38/38 - 4s - 97ms/step - accuracy: 0.6440 - loss: nan
Epoch 2/44
38/38 - 1s - 17ms/step - accuracy: 0.6440 - loss: nan
Epoch 3/44
38/38 - 1s - 17ms/step - accuracy: 0.6440 - loss: nan
Epoch 4/44
38/38 - 1s - 33ms/step - accuracy: 0.6440 - loss: nan
Epoch 5/44
38/38 - 1s - 18ms/step - accuracy: 0.6440 - loss: nan
Epoch 6/44
38/38 - 1s - 17ms/step - accuracy: 0.6440 - loss: nan
Epoch 7/44
38/38 - 1s - 33ms/step - accuracy: 0.6440 - loss: nan
Epoch 8/44
38/38 - 1s - 33ms/step - accuracy: 0.6440 - loss: nan
Epoch 9/44
38/38 - 1s - 32ms/step - accuracy: 0.6440 - loss: nan
Epoch 10/44
38/38 - 1s - 18ms/step - accuracy: 0.6440 - loss: nan
Epoch 11/44
38/38 - 1s - 17ms/step - accuracy: 0.6440 - loss: nan
Epoch 12/44
38/38 - 2s - 44ms/step - accuracy: 0.6440 - loss: nan
Epoch 13/44
38/38 - 1s - 33ms/step - accuracy: 

/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/44


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequ

Epoch 1/44
Epoch 1/44


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


| 2         | nan       | 7.492     | 369.9     | 0.1818    | 0.05502   | 44.34     | 2.05      | 1.864     | 1.582     | 0.6157    | 22.55     | 0.2921    | 2.565     |
Epoch 1/67


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 770, in fit
    self._fit(
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrapp

17/17 - 8s - 454ms/step - accuracy: 0.3800 - loss: 15.8136
Epoch 2/67
17/17 - 3s - 202ms/step - accuracy: 0.5165 - loss: 15.7736
Epoch 3/67
17/17 - 3s - 184ms/step - accuracy: 0.5292 - loss: 9.3243
Epoch 4/67
17/17 - 3s - 190ms/step - accuracy: 0.5269 - loss: 8.8360
Epoch 5/67
17/17 - 7s - 386ms/step - accuracy: 0.5326 - loss: 8.2870
Epoch 6/67
17/17 - 4s - 220ms/step - accuracy: 0.5316 - loss: 9.8553
Epoch 7/67
17/17 - 6s - 376ms/step - accuracy: 0.4833 - loss: 11.3362
Epoch 8/67
17/17 - 4s - 210ms/step - accuracy: 0.5283 - loss: 14.0525
Epoch 9/67
17/17 - 5s - 304ms/step - accuracy: 0.5313 - loss: 13.2863
Epoch 10/67
17/17 - 6s - 380ms/step - accuracy: 0.5438 - loss: 10.4571
Epoch 11/67
17/17 - 4s - 225ms/step - accuracy: 0.5192 - loss: 11.7542
Epoch 12/67
17/17 - 3s - 180ms/step - accuracy: 0.5352 - loss: 9.1685
Epoch 13/67
17/17 - 4s - 211ms/step - accuracy: 0.5371 - loss: 8.8538
Epoch 14/67
17/17 - 4s - 255ms/step - accuracy: 0.5454 - loss: 8.0969
Epoch 15/67
17/17 - 4s - 229ms/st

/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/67


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/67


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/67


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


| 3         | nan       | 4.105     | 828.1     | 0.1997    | 0.1543    | 67.39     | 1.093     | 2.215     | 1.341     | 0.0744    | 95.4      | 0.9656    | 5.659     |
Epoch 1/30


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 770, in fit
    self._fit(
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrapp

50/50 - 4s - 83ms/step - accuracy: 0.4330 - loss: 9.6512
Epoch 2/30
50/50 - 4s - 87ms/step - accuracy: 0.5101 - loss: 5.7256
Epoch 3/30
50/50 - 2s - 43ms/step - accuracy: 0.5254 - loss: 3.7886
Epoch 4/30
50/50 - 2s - 43ms/step - accuracy: 0.5295 - loss: 2.8082
Epoch 5/30
50/50 - 2s - 43ms/step - accuracy: 0.5822 - loss: 1.4527
Epoch 6/30
50/50 - 4s - 79ms/step - accuracy: 0.6021 - loss: 1.1518
Epoch 7/30
50/50 - 6s - 114ms/step - accuracy: 0.6125 - loss: 1.0739
Epoch 8/30
50/50 - 7s - 133ms/step - accuracy: 0.6365 - loss: 0.9958
Epoch 9/30
50/50 - 3s - 69ms/step - accuracy: 0.6211 - loss: 1.0417
Epoch 10/30
50/50 - 4s - 76ms/step - accuracy: 0.6600 - loss: 0.9467
Epoch 11/30
50/50 - 2s - 42ms/step - accuracy: 0.6783 - loss: 0.9136
Epoch 12/30
50/50 - 2s - 42ms/step - accuracy: 0.6886 - loss: 0.8776
Epoch 13/30
50/50 - 2s - 44ms/step - accuracy: 0.5947 - loss: 1.3651
Epoch 14/30
50/50 - 3s - 67ms/step - accuracy: 0.6974 - loss: 0.8791
Epoch 15/30
50/50 - 4s - 78ms/step - accuracy: 0.692

13/13 - 0s - 27ms/step
Epoch 1/30


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/30


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/30


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/30


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


| 4         | nan       | 2.742     | 278.1     | 0.6842    | 0.132     | 29.76     | 1.99      | 1.069     | 2.819     | 0.2662    | 69.63     | 0.3117    | 3.64      |
Epoch 1/95


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 770, in fit
    self._fit(
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrapp

40/40 - 2s - 61ms/step - accuracy: 0.4607 - loss: 8422.8340
Epoch 2/95
40/40 - 1s - 32ms/step - accuracy: 0.4857 - loss: 28.5549
Epoch 3/95
40/40 - 1s - 20ms/step - accuracy: 0.3836 - loss: 193.9839
Epoch 4/95
40/40 - 2s - 40ms/step - accuracy: 0.4471 - loss: 376.6125
Epoch 5/95
40/40 - 1s - 37ms/step - accuracy: 0.4599 - loss: 17.0991
Epoch 6/95
40/40 - 2s - 51ms/step - accuracy: 0.4540 - loss: 20.0196
Epoch 7/95
40/40 - 1s - 31ms/step - accuracy: 0.4369 - loss: 21.9476
Epoch 8/95
40/40 - 1s - 32ms/step - accuracy: 0.4112 - loss: 22.4917
Epoch 9/95
40/40 - 1s - 32ms/step - accuracy: 0.4262 - loss: 22.8979
Epoch 10/95
40/40 - 1s - 32ms/step - accuracy: 0.4143 - loss: 32.0130
Epoch 11/95
40/40 - 1s - 21ms/step - accuracy: 0.4244 - loss: 23.7877
Epoch 12/95
40/40 - 1s - 31ms/step - accuracy: 0.4270 - loss: 22.6952
Epoch 13/95
40/40 - 1s - 31ms/step - accuracy: 0.4050 - loss: 23.3819
Epoch 14/95
40/40 - 1s - 22ms/step - accuracy: 0.4190 - loss: 23.2968
Epoch 15/95
40/40 - 2s - 43ms/step -

/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/95


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/95


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/95


/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


| 5         | nan       | 4.92      | 347.9     | 0.9696    | 0.2325    | 95.16     | 2.79      | 2.196     | 2.844     | 0.09761   | 27.64     | 0.04523   | 2.277     |


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:547: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
4 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 895, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrappers.py", line 770, in fit
    self._fit(
  File "/usr/local/lib/python3.11/dist-packages/scikeras/wrapp

ValueError: Input y contains NaN.

In [39]:
optimum = nn_opt.max['params']
learning_rate = optimum['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU, 'relu']
optimum['activation'] = activationL[round(optimum['activation'])]
optimum['batch_size'] = round(optimum['batch_size'])
optimum['epochs'] = round(optimum['epochs'])
optimum['layers1'] = round(optimum['layers1'])
optimum['layers2'] = round(optimum['layers2'])
optimum['neurons'] = round(optimum['neurons'])
optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(learning_rate=learning_rate),
             'SGD':SGD(learning_rate=learning_rate),
             'RMSprop':RMSprop(learning_rate=learning_rate),
             'Adadelta':Adadelta(learning_rate=learning_rate),
             'Adagrad':Adagrad(learning_rate=learning_rate),
             'Adamax':Adamax(learning_rate=learning_rate),
             'Nadam':Nadam(learning_rate=learning_rate),
             'Ftrl':Ftrl(learning_rate=learning_rate)}
optimum['optimizer'] = optimizerD[optimizerL[round(optimum['optimizer'])]]
optimum

{'activation': 'softsign',
 'batch_size': 961,
 'dropout': 0.7319939418114051,
 'dropout_rate': 0.17959754525911098,
 'epochs': 32,
 'kernel': 1.3119890406724053,
 'layers1': 1,
 'layers2': 3,
 'learning_rate': 0.6051038616257767,
 'neurons': 74,
 'normalization': 0.020584494295802447,
 'optimizer': <keras.src.optimizers.ftrl.Ftrl at 0x7d2df6121050>}

In [40]:
optimum = {'activation': 'softsign',
 'batch_size': 961,
 'dropout': 0.7319939418114051,
 'dropout_rate': 0.17959754525911098,
 'epochs': 32,
 'kernel': 1.3119890406724053,
 'layers1': 1,
 'layers2': 3,
 'learning_rate': 0.6051038616257767,
 'neurons': 74,
 'normalization': 0.020584494295802447,
 'optimizer': 'Adadelta'
}

In [41]:

epochs = 32
batch_size = 961
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = 15
layers1 = 1
layers2 = 3
activation = 'softsign'
kernel = 1
neurons = 74
normalization = 0.02058
dropout = 0.7320
dropout_rate = 0.1796
optimizer = 'Adadelta'
learning_rate = 0.6051038616257767

In [43]:
model = Sequential()
model.add(Conv1D(neurons, kernel_size=kernel, activation=activation, input_shape=(timesteps, input_dim)))
if normalization > 0.5:
    model.add(BatchNormalization())
for i in range(layers1):
    model.add(Dense(neurons, activation=activation))
if dropout > 0.5:
    model.add(Dropout(dropout_rate, seed=123))
for i in range(layers2):
    model.add(Dense(neurons, activation=activation))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax'))
optimizer_instance = Adadelta(learning_rate=learning_rate)
model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer_instance, metrics=['accuracy'])
model.summary()

Model: "sequential_26"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv1d_26 (Conv1D)                   │ (None, 15, 74)              │             740 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_130 (Dense)                    │ (None, 15, 74)              │           5,550 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_16 (Dropout)                 │ (None, 15, 74)              │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_131 (Dense)                    │ (None, 15, 74)              │           5,550 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_132 (Dense)                    │ (None, 15, 74)              │           5,550 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_133 (Dense)                    │ (None, 15, 74)              │           5,550 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling1d_26 (MaxPooling1D)      │ (None, 7, 74)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_26 (Flatten)                 │ (None, 518)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_134 (Dense)                    │ (None, 15)                  │           7,785 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 30,725 (120.02 KB)

 Trainable params: 30,725 (120.02 KB)

 Non-trainable params: 0 (0.00 B)

In [44]:
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

Epoch 1/32
18/18 - 6s - 334ms/step - accuracy: 0.5747 - loss: 1.6153
Epoch 2/32
18/18 - 6s - 339ms/step - accuracy: 0.6451 - loss: 1.0908
Epoch 3/32
18/18 - 4s - 212ms/step - accuracy: 0.6476 - loss: 1.0497
Epoch 4/32
18/18 - 3s - 190ms/step - accuracy: 0.6465 - loss: 1.0367
Epoch 5/32
18/18 - 6s - 349ms/step - accuracy: 0.6418 - loss: 1.0235
Epoch 6/32
18/18 - 4s - 217ms/step - accuracy: 0.6472 - loss: 1.0064
Epoch 7/32
18/18 - 5s - 276ms/step - accuracy: 0.6541 - loss: 0.9826
Epoch 8/32
18/18 - 7s - 365ms/step - accuracy: 0.6585 - loss: 0.9629
Epoch 9/32
18/18 - 4s - 196ms/step - accuracy: 0.6631 - loss: 0.9434
Epoch 10/32
18/18 - 7s - 393ms/step - accuracy: 0.6730 - loss: 0.9149
Epoch 11/32
18/18 - 4s - 218ms/step - accuracy: 0.6938 - loss: 0.8729
Epoch 12/32
18/18 - 4s - 248ms/step - accuracy: 0.7007 - loss: 0.8533
Epoch 13/32
18/18 - 5s - 286ms/step - accuracy: 0.7076 - loss: 0.8339
Epoch 14/32
18/18 - 5s - 272ms/step - accuracy: 0.7231 - loss: 0.8075
Epoch 15/32
18/18 - 3s - 181m

In [45]:
predictions = {
0: 'BASEL',
1: 'BELGRADE',
2: 'BUDAPEST',
3: 'DEBILT',
4: 'DUSSELDORF',
5: 'HEATHROW',
6: 'KASSEL',
7: 'LJUBLJANA',
8: 'MAASTRICHT',
9: 'MADRID',
10: 'MUNCHENB',
11: 'OSLO',
12: 'SONNBLICK',
13: 'STOCKHOLM',
14: 'VALENTIA'
}

In [46]:
y_train.shape

(17212,)

In [47]:
def confusion_matrix(y_true, y_pred, predictions):
    # Check if y_true and y_pred are one-hot encoded or already class indices
    if y_true.ndim == 1:
        y_true_labels = y_true
    else:
        y_true_labels = np.argmax(y_true, axis=1)

    if y_pred.ndim == 1:
        y_pred_labels = y_pred
    else:
        y_pred_labels = np.argmax(y_pred, axis=1)

    # Map numeric labels to activity names
    y_true_series = pd.Series([predictions[y] for y in y_true_labels])
    y_pred_series = pd.Series([predictions[y] for y in y_pred_labels])

    return pd.crosstab(y_true_series, y_pred_series, rownames=['True'], colnames=['Pred'])

In [48]:
pred = model.predict(X_test)
pred.shape

180/180 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step


(5738, 15)

In [49]:
y_test.shape

(5738,)

In [50]:
ty= to_categorical(y_test)
print(ty)

[[1. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 ...
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]


In [51]:
# Evaluate
print(confusion_matrix(y_test, pred, predictions))

Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  HEATHROW  LJUBLJANA  MADRID
True                                                                      
BASEL        3221       328        32      15        14          5      67
BELGRADE      212       817        19       3         3          1      37
BUDAPEST       45        66        68       5         4          1      25
DEBILT         14        18        11      23         8          1       7
DUSSELDORF      7         6         6       2         3          1       4
HEATHROW       22         6         1       2        15          1      35
KASSEL          6         2         2       0         1          0       0
LJUBLJANA      19         8         4       0         0          3      27
MAASTRICHT      9         0         0       0         0          0       0
MADRID        112        48         9       0         2          1     286
MUNCHENB        8         0         0       0         0          0       0
OSLO            3        